### ISSUE TITLE
Consider updating IdentityUserClaim to store the claim value type

### ISSUE THREAD

**Issue by PinpointTownes** [aspnet/Identity#1245]
> Be aware that ASP.NET Identity doesn’t store claim value types, so even in cases where the claim is always an integer (as in this example), it will be stored and returned as a string. Later in this post, I explain how non-string claims can be included in JWT tokens.

https://blogs.msdn.microsoft.com/webdev/2016/10/27/bearer-token-authentication-in-asp-net-core/

**Comment by PinpointTownes** [comment]
In the same vein, I suppose the issuer could also be stored.

**Comment by blowdart** [comment]
Hmm limited to jwt? An issue for identity model, not us I'm afraid.

**Comment by PinpointTownes** [comment]
@blowdart heh, it's not limited to JWT at all (the blog post just illustrates a use case where it would be useful).

And BTW, IdentityModel already does the right thing as it serializes claims correctly according to their value type. It's just Identity that doesn't preserve/store the value type when adding a claim via `UserManager.AddClaimAsync()`.

**Comment by blowdart** [comment]
Oh I see. Hmm, bother. @HaoK ... this would take a flipping huge database change :(

**Comment by PinpointTownes** [comment]
Luckily, you now have a *super* versioning strategy to handle these changes :trollface:

**Comment by blowdart** [comment]
As you're well aware we're already doing a bunch of horrid changes to the database already for 2.0. Adding more at this point in time would add to risk. Unless you have a business requirement because of this, rather than "it'd be nice"

**Comment by PinpointTownes** [comment]
I'm not sure I understand how this would add to risk, since adding properties to existing Identity models is already something you do in 2.0: https://github.com/aspnet/Identity/blob/dev/src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/IdentityDbContext.cs#L246-L248

**Comment by blowdart** [comment]
Yes, and every time we do it, we don't like it, as a failed migration may end up in a horrible state.

**Comment by blowdart** [comment]
So is there a reason other than it'd be nice? That would help justify the risk.

**Comment by PinpointTownes** [comment]
It's an annoying limitation that requires using nasty hacks such as removing/adding claims manually as explained in the blog post, so yes, it would be "nice" to remove it. Now, it's up to you to determine whether it should be considered as a bug/oversight that should be fixed or just a welcome improvement.

**Comment by blowdart** [comment]
We'll triage it at the weekly meeting.

**Comment by PinpointTownes** [comment]
Thanks Barry :four_leaf_clover:

**Comment by HaoK** [comment]
This one shouldn't be too bad, it'd be adding two new string properties to the Identity[User/Role]Claim tables, Issuer/ClaimType.  We already take Claims as the inputs on the store, so it doesn't actually affect any of the public APIs, it'd just be an additional switch we'd need to have inside of the store implementations to check the StoreVersion to control whether we use the new fields.

**Comment by blowdart** [comment]
OK this is dependent on time remaining which we're not sure about yet. If we have time, then yes, along with OriginalIssuer too.

**Comment by HaoK** [comment]
@ajcvickers feature?

**Comment by HaoK** [comment]
If push comes to shove for scheduling, this is something we could do in 2.1 anyways...

**Comment by ajcvickers** [comment]
Yes, should probably punt out to 2.1.

**Comment by HaoK** [comment]
@blowdart @ajcvickers we should move this out if we aren't planning on making any schema changes for Identity in 2.1
